### PR TITLE
Failsafe - Allow re-arming after failsafe condition.

### PR DESCRIPTION
**Pull request by ctzsnooze**
With current code, if an aircraft goes through the CleanFlight Failsafe landing phase and enters 'Landed', there is a lock set on re-arming.  It is impossible to re-arm, even if the radio link is restored, without physically power cycling the aircraft.

This pull request allows re-arming after failsafe.  I do not propose a CLI setting for this, because I just want to re-arm at will  :-) but others may want a CLI entry - say allow_rearm_after_landing where 0 is current functionality 1 is you can rearm if you have signal.

The intent appears to have been based on a concern that intermittent signal could have unpredictable results.  

However, CleanFlight is quite good at detecting normal incoming signals, many of us have RSSI so we know we can talk reliably to the Rx, and there are situations where being able to re-arm without power cycling would be an advantage.

For instance:

- a quadcopter goes past signal range, you can get close enough to see it can arm safely, but it is still a long walk away, and (maybe) you have RSSI back to know Rx signal strength.  It might be nice to re-arm and fly back without walking or climbing all the way to the quad.  And indeed the quad may be inaccessible, behind a fence, or otherwise up a tree or difficult to access with no other possibility but to fly out.

-  we take off but have forgotten to put our Tx antenna on, or it was broken, quad goes failsafe and lands but is 200m away.  We replace antenna, arm, fly back - or we walk 200m.  Mostly we want just to arm and fly back.

- we are flying and go failsafe because signal is blocked by a nearby object.  We walk around the object and get good signal back, but cannot re-arm.  Quad is a long way away... :-)

I have not had issues with false re-arming since disabling this lock.  I think that's because CF is very good at cleaning bad PPM pulses.  The recent changes to rxfail and pulse checking make the original need to stop re-arming less of an issue, I think.

I hope this can go into the master, probably with a CLI control line to preserve current behaviour.

Discussed here #995 and briefly some time back here  #744

Please vote here for your preferred behaviour?

### CODE REVIEW

**Comment by Redshifft** [comment]
I just double checked, It's defiantly not the first of April, I really can not believe some of what's been posted on failsafes  today.
If you operate a Quad it's your responsibility to know the limitations of your control equipment and if you get caught out the walk will be a timely reminder !

**Comment by ctzsnooze** [comment]
If the CF failsafe is not enabled, and the pilot uses only standard Rx failsafe, as many do, then the behaviour of the system is as I describe above.  Tell me one Rx that locks you out of the chance to re-connect, permanently, after loss of signal (regardless of duration)?  So why is it so good that Cleanflight locks you out?

**Comment by Redshifft** [comment]
@ctzsnooze  What do you want on your headstone ? :D 

The top RC PCM systems before 2.4ghz had failsafes that were not reset until they were power cycled, most had a delay setting you could adjust before it fully failsafed.
Failsafes generally always override and lock down systems to prevent damage etc in all types of equipment, and if it did not stay failsafed the source of the problem could cause something really nasty.

With the advent of 2.4ghz and it's delicate LOS signal also came better micro processors, the systems can tolerate quite a bit of packet loss before deciding it's an unrecoverable situation and dropping to failsafe and like you say are happy to continue when the link is restored

Problem we have with a Quad is we don't actually fly it, we guide it with our RC, the FC does the flying, Logically it follows that if there is a problem down stream of the FC, the flight controller should be able to step in and go into some sort of Failsafe routine, because the RC is totally irrelevant if it malfunctions and decides it wants to keep flying 

With little miniquads, just kill the power and make them safe, they can all take a hit, I just have not seen any that can fly back out from trees or crashes the other side of walls
So running just a Radio failsafe is better than nothing but really not that good.

If You get a glitch or lockout, immediately hold up your TX, the better signal propagation should save you..land and investigate what was blanking the signal. If it's not your problem it could be the general "noise" floor causing the RX to back down it's front end sensitivity and in so doing limiting your range.

Bottom line, like that Airbag in your car you should never use it but that little light still comes on with the ignition to tell you it's got your back if things go bad.

**Comment by ethomas997** [comment]
Personally, I'm fine with getting rid of the arming-after-failsafe block.  For AUX-switch arming It would be good to have a requirement that the switch be in the disarm position for a length of time before arming is allowed, but some version of that is probably in the code already.

If there are qualms about safety, another option could be to have a special action to allow the re-arm.  If stick arming is in use, the action could be to hold the stick in the disarm position for a certain time period, and then regular stick arming would be allowed.  If AUX-switch arming is in use, the action could be to toggle the switch a number of times during a certain time period.

In PR #995 Hydra mentioned having the copter stuck in a tree as a possible reason for wanting to always be able to rearm.  Very often in that case the rearm will not be possible because the copter is at too great of an angle (and 'small_angle' == something like 25).  It would be good if an action like the above or similar could be an override so as to always have a way to arm the copter in a pinch.

--ET

**Comment by henn1001** [comment]
I´m agreeing on this override feature aswell - i can´t think of a scenario where this is bad? can´t understand that fear mongering from Redshift..

**Comment by borisbstyle** [comment]
@ethomas997 When your copter is in tree outside small angle you can recalibrate ACC on sticks and you will be able to arm again ;)

**Comment by Redshifft** [comment]
@henn1001  my sentiments are for the great unwashed, to protect others
I'm the first to say we should all have what we want and anyone who has contributed around here should be smart enough to not abuse it if we simply add an option in the CLI to override the default off...don't you agree ? It is after all a safety feature and not even something we should ever needing to be getting around regularly.

**Comment by henn1001** [comment]
Yeah of course - this pull-request is too radical ofc. thats why i think we need some kind of mechanism to override this.

**Comment by ctzsnooze** [comment]
Probably the majority of CleanFlight users set their Rx to send custom values on failsafe, or leave the default of hold last values, rather than use the internal CleanFlight failsafe.  

This way they never get the 1500 issue, they get their control back immediately signal is restored, they can arm again anytime they like.  

Exactly why anyone would *not* want to get control back immediately signal is restored - whether before failsafe is active or after - beats me.  

Lots of times you get glitching and restoring and it doesn't of itself cause bad things.  Lots of people using Rx based failsafe lose signal and get it back without problems.  Cleanflight allows immediate return of control even after quite long periods of time (failsafe delay time + failsafe time).  But all of a sudden, once it has 'landed', you are locked out.  This just makes no sense to me.

So far no-one here has given me an example of an actual event where getting control back was a problem.  People say the data may be unreliable or bad things might happen, but it doesn't.  

The only reasonable explanation is that it is protecting the developers from legal liability somehow.

**Comment by leothehuman** [comment]
First, you should have also updated the test: cleanflight/src/test/unit/flight_failsafe_unittest.cc
Second, I would prefer having to disarm the quad before arming it again. It is quite likely to have the arm switch in the same armed position and bump the throttle while you're still searching for a quad. It's also quite easy to disarm the quad before arming it again after a failsafe has kicked in. I would disable the PREVENT_ARMING flag on successful disarm (from either armed or failsafe state) instead of not setting it at all.

**Comment by ethomas997** [comment]
The thinking / worry was probably that if the RX signal has turned into noise on a copter with AUX-switch arming, the noise could be interpreted as arm-AUX-switch 'on' and throttle up.  If the failsafe condition lasts long enough to get through the auto-landing sequence, then the RX signal is considered "bad" and rearming is disabled.

If the code has checks that make sure the copter has been disarmed for a while before arming and to make sure that the arming-channel signal is stable, then that's probably plenty safe enough.  The copter is always automatically disarmed after a failsafe auto-landing (regardless of arm-AUX-switch position), and that behavior should of course be retained.  (There's also the option of implementing the "special actions" I mentioned above.)

@borisbstyle That's a good trick, redoing the ACC calibration -- as long as one remembers the stick sequence at the time (or has access to the doc).

--ET

**Comment by ctzsnooze** [comment]
If someone could guide me how to make the re-disarm happen before arming, that would be great.  

In testing I've not encountered any problems so far.  

The problem concerning people here is if someone sought to re-arm when signal was marginal.  That's not a great idea for many reasons.  But people do this.  The main thing is to provide the option.  If you enable that option you will find, on the rare occasion that you have a failsafe event, that it is useful and makes for a better day.

**Comment by leothehuman** [comment]
That's quite easy. In all places, where disarm conditions are checked and verified, you need to make sure to clear the PREVENT_ARMING flag, regardless of the arm status itself (like, if I disarm the quad, that is disarmed by the failsafe, it should still reset the flag). You may need to move the disarming code to a place common to the armed and disarmed state to do that.

**Comment by ctzsnooze** [comment]
That coding may or may not be easy; it is beyond me, I think!  :-)

The problem of self-re-arming when signal is marginal is, I think, greatly over-rated.  

There is smoothing and error detection for RC inputs, and it takes valid data to arm that just doesn't seem to happen unless the signal is quite robust.  

I have a dodgy spektrum sat Rx that I test this kind of stuff with.  I've found that if you can arm after a failsafe, even at the point where you are only just able to arm and there is still glutting, false arming just does not occur.  To date I've never had the quad self-arm while disabled in the presence of marginal signal (or no signal).  At least that's my experience to date.

It is easy to test this yourself, because anytime the quad is powered up and the Rx signal is marginal, this risk exists.  So just set your Tx/Rx in range test mode, arm, walk away while wobbling it on the ground until signal is marginal or you get gitching or you start getting close to failsafe, disarm... and wait for it to self-arm... it just doesn't happen.  Thank goodness for Rx and CF functionality to remove false data, I guess!  :-)

Those of us who are concerned about the risks, for a particular quad, do not need to enable this option if it existed - and it should be off by default.  

I have noticed that at some times a manual disarm is required before re-arming is permitted.  So your proposal is definitely possible - I just am not sure at all if I can implement it.

**Comment by ctzsnooze** [comment]
Disarm is ultimately handled in mw.c around lines 314:

    void mwDisarm(void)
    {
        if (ARMING_FLAG(ARMED)) {
            DISABLE_ARMING_FLAG(ARMED);
    
    #ifdef BLACKBOX
            if (feature(FEATURE_BLACKBOX)) {
                finishBlackbox();
            }
    #endif
    
            beeper(BEEPER_DISARMING);      // emit disarm tone
        }
    }

Maybe, instead of my original proposal, the following would achieve the clear of the block - but only after a successful disarm:

    void mwDisarm(void)
    {
    if (ARMING_FLAG(PREVENT_ARMING)) {
        DISABLE_ARMING_FLAG(PREVENT_ARMING);
    }
    if (ARMING_FLAG(ARMED)) {
        DISABLE_ARMING_FLAG(ARMED);   <..... then the code to stop blackbox and beep>

I think that with the above, the only way to clear the prevent_arming flag would be to first achieve successful disarm.

Would this be better?

**Comment by leothehuman** [comment]
Yep, looks exactly like what I proposed. I'm not sure if the "if (ARMING_FLAG(PREVENT_ARMING))" is necessary there, we can just clear the flag either way.

**Comment by leothehuman** [comment]
If you have already made the changes, or plan to in the nearest future, please, update the PR. Otherwise, I will be coding this over the weekend.

**Comment by hydra** [comment]
@ctzsnooze Thanks for addressing peoples concerns, yes requiring a disarm seems quite a sensible approach.  I will look into this for a post 1.10.0 release.

**Comment by leothehuman** [comment]
Is there any chance we can integrate this 3 liner earlier? It doesn't look too dangerous.

**Comment by ProDrone** [comment]
I added this feature in a different way to the failsafe handling features in #852
Read more in the applicable commit message or the first message on top.

**Comment by ProDrone** [comment]
This is now fully integrated and tested in #852, also the unitttests are adapted.
@ctzsnooze So maybe you want to close this to reduce the number of open PR's

**Comment by ctzsnooze** [comment]
Yes, Thanks ProDrone, I'll close this.  I really appreciate all the work you've done!  :-)
